mlflow/types/schema.py

```diff
@@ -1329,3 +1330,116 @@ def __eq__(self, other) -> bool:
 
     def __repr__(self) -> str:
         return repr(self.params)
+
+
+def convert_dataclass_to_schema(dataclass):
+    """
+    Create a Schema object from a dataclass. The dataclass should have type hints
+    for each field. Fields in the dataclass can be composed of basic types or other dataclasses.
+    Only the top-level fields become ColSpecs directly. Lower-level fields are converted into
+    nested Object types.
+    """
+    field_type_mapping = {
+        str: "string",
+        int: "integer",
+        bool: "boolean",
+    }
+
+    inputs = []
+
+    for field_name, field_type in dataclass.__annotations__.items():
+        # Determine the type and handle Optional and List correctly
+        is_optional = False
+        effective_type = field_type
+
+        if get_origin(field_type) == Union and type(None) in get_args(field_type):
+            # This is an Optional type; determine the effective type excluding None
+            is_optional = True
+            effective_type = next(t for t in get_args(field_type) if t is not type(None))
+
+        if get_origin(effective_type) == list:
+            # It's a list, check the type within the list
+            list_type = get_args(effective_type)[0]
+            if is_dataclass(list_type):
+                dtype = convert_dataclass_to_nested_object(list_type)  # Convert to nested Object
+                inputs.append(
+                    ColSpec(type=Array(dtype=dtype), name=field_name, required=not is_optional)
+                )
+            else:
+                inputs.append(
+                    ColSpec(
+                        type=Array(dtype=field_type_mapping[list_type]),
+                        name=field_name,
+                        required=not is_optional,
+                    )
+                )
+        elif is_dataclass(effective_type):
+            # It's a nested dataclass
+            dtype = convert_dataclass_to_nested_object(effective_type)  # Convert to nested Object
+            inputs.append(
+                ColSpec(
+                    type=dtype,
+                    name=field_name,
+                    required=not is_optional,
+                )
+            )
+        else:
+            # It's a basic type
+            inputs.append(
+                ColSpec(
+                    type=field_type_mapping[effective_type],
+                    name=field_name,
+                    required=not is_optional,
+                )
+            )
+
+    return Schema(inputs=inputs)
+
+
+def convert_dataclass_to_nested_object(dataclass):
+    """
+    Convert a nested dataclass to an Object type used within a ColSpec.
+    """
+    properties = []
+    for field_name, field_type in dataclass.__annotations__.items():
+        properties.append(convert_field_to_property(field_name, field_type))
+    return Object(properties=properties)
+
+
+def convert_field_to_property(field_name, field_type):
```

Comment:
I think we can refactor this function with convert_dataclass_to_schema together, make this a function that returns dtype (instead of Property) and use in both places

Reply:
Not a blocker!